examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2218 @@
+  //    <code>copy_local_to_global</code> in charge of actually coping these
+  //    local contributions into the global objects (matrices and/or vectors)
+  //
+  // Most the following lines are spent in the definition of the worker
+  // <code>local_assemble_system</code> and the copy data routine
+  // <code>copy_local_to_global</code>. There is not much to say about the
+  // WorkStream framework since the vast majority of ideas are reasonably
+  // well-documented in Step-9, Step-13 and Step-32 among others.
+  //
+  // Finally the boundary normals are defined as
+  // $\widehat{\boldsymbol{\nu}}_i =
+  // \frac{\boldsymbol{\nu}_i}{|\boldsymbol{\nu}_i|}$ where
+  // $\boldsymbol{\nu}_i = \sum_{F \subset \text{supp}(\phi_i)}
+  // \sum_{\mathbf{x}_{q,F}} \nu(\mathbf{x}_{q,F})
+  // \phi_i(\mathbf{x}_{q,F})$, here: $F \subset \partial \Omega$ denotes 
+  // faces of elements at the boundary of the domain, and $\mathbf{x}_{q,F}$ 
+  // are quadrature points on such face.
+  // Other more sophisticated definitions for $\nu_i$ are 
+  // possible but none of them have much influence in theory or practice.
+  // We remind the reader that <code>CopyData</code> includes the class member
+  // <code>local_boundary_normal_map</code> in order to store these local
+  // contributions for the boundary map.
+
+  template <int dim>
+  void OfflineData<dim>::assemble()
+  {
+    lumped_mass_matrix = 0.;
+    norm_matrix        = 0.;
+    for (auto &matrix : cij_matrix)
+      matrix = 0.;
+    for (auto &matrix : nij_matrix)
+      matrix = 0.;
+
+    const unsigned int dofs_per_cell =
+      discretization->finite_element.dofs_per_cell;
+    const unsigned int n_q_points = discretization->quadrature.size();
+
+    /* This is the implementation of the scratch data required by WorkStream */
+    MeshWorker::ScratchData<dim> scratch_data(
+      discretization->mapping,
+      discretization->finite_element,
+      discretization->quadrature,
+      update_values | update_gradients | update_quadrature_points |
+        update_JxW_values,
+      discretization->face_quadrature,
+      update_normal_vectors | update_values | update_JxW_values);
+
+    {
+      TimerOutput::Scope t(
+        computing_timer,
+        "offline_data - assemble lumped mass matrix, and c_ij");
+
+      /* This is the implementation of the "worker" required by WorkStream */
+      const auto local_assemble_system = [&](const auto &cell,
+                                             auto &      scratch,
+                                             auto &      copy) {
+        auto &is_artificial             = copy.is_artificial;
+        auto &local_dof_indices         = copy.local_dof_indices;
+        auto &local_boundary_normal_map = copy.local_boundary_normal_map;
+        auto &cell_lumped_mass_matrix   = copy.cell_lumped_mass_matrix;
+        auto &cell_cij_matrix           = copy.cell_cij_matrix;
+
+        is_artificial = cell->is_artificial();
```

Comment:
why do you run on artificial cells? why this second assignment (see 6 lines above)?

Reply:
We are iterating over locally owned cells and the ghost layer, i.e., all cells that are not artificial. The first statement creates a reference, the second statement assigns the value.

Reply:
@tjhei's confusion illustrates a broader point: Introducing alias variables is generally difficult to read. I would encourage just using the original, unaliased set of variables even though that takes 5 more characters to write. (But you get to save yourself the declarations of the alias variables.)